toolchain/check/convert.cpp

```diff
@@ -1059,25 +1062,55 @@ static auto PerformBuiltinConversion(Context& context, SemIR::LocId loc_id,
       }
     }
 
-    if (sem_ir.types().Is<SemIR::FacetType>(
-            sem_ir.insts().Get(lookup_inst_id).type_id())) {
-      // Conversion from a facet value to a `FacetValue` of the target
-      // `FacetType`. We move up to the higher typish level, and convert the
-      // `FacetType` (which is of type TypeType) directly. If the `FacetType`
-      // itself implements the target `FacetType` (`impl Iface1 as Iface2`),
-      // then we will produce a `FacetValue` of the target `FacetType`.
-      lookup_inst_id = context.types().GetInstId(
-          sem_ir.insts().Get(lookup_inst_id).type_id());
+    // Conversion from a facet value (which has type `FacetType`) or a type
+    // value (which has type `TypeType`) to a facet value. We can do this if the
+    // type satisfies the requirements of the target `FacetType`, as determined
+    // by finding impl witnesses for the target FacetType.
+    //
+    // This binds the value to the `FacetType` with a `FacetValue`, which
+    // requires an instruction of type `TypeType`. So if we are converting from
+    // a facet value, we get its `type` via an extra `FacetAccessType`
+    // instruction.
+    auto lookup_result = LookupImplWitness(
+        context, loc_id, sem_ir.constant_values().Get(lookup_inst_id),
+        sem_ir.types().GetConstantId(target.type_id));
+    if (lookup_result.has_value()) {
+      if (lookup_result.has_error_value()) {
+        return SemIR::ErrorInst::SingletonInstId;
+      } else {
+        auto type_inst_id = SemIR::InstId::None;
+        if (sem_ir.types().Is<SemIR::FacetType>(lookup_inst_type_id)) {
+          type_inst_id =
+              AddInst(context, loc_id,
+                      SemIR::FacetAccessType{
+                          .type_id = SemIR::TypeType::SingletonTypeId,
+                          .facet_value_inst_id = lookup_inst_id,
+                      });
+        } else {
+          type_inst_id = lookup_inst_id;
+        }
+        return AddInst<SemIR::FacetValue>(
+            context, loc_id,
+            {.type_id = target.type_id,
+             .type_inst_id = type_inst_id,
+             .witnesses_block_id = lookup_result.inst_block_id()});
+      }
     }
```

Comment:
Again, if this failed, I don't think we should be trying to call `LookupImplWitness` again.

Reply:
Done